### DISCUSSION TITLE
Emojis not available in ChatResponse content

### DISCUSSION THREAD

**Discussion by grzegorz-aniol** [spring-projects/spring-ai#1288] [Q&A]
Hi, I'm testing a prompt that generates text enriched with emojis. I'm using OpenAI. It works well in the ChatGPT console; however, I don't see emojis in the `ChatResponse` content. Does this mean the implementation should encode JSON with `APPLICATION_JSON_UTF8`? I tried to add my own codec to `WebClient`, but it doesn't work. Has anyone had a similar problem or knows how to solve it? Thanks! :)

**Comment by RodrigoDiasDeOliveira** [comment]
It sounds like you're running into an issue where the emoji characters are not being correctly rendered or processed in the ChatResponse content of your implementation. This is a common issue when dealing with text encoding, especially with special characters like emojis.

Here are a few things you can check or try to solve this issue:

1. Ensure Correct Encoding in the API Response:
When using APPLICATION_JSON_UTF8, it’s important to ensure that your response content is correctly encoded. The standard APPLICATION_JSON media type is generally used with UTF-8 encoding by default, so explicitly setting APPLICATION_JSON_UTF8 may not always be necessary.
However, ensure that your server and client are both correctly handling UTF-8 encoding. Double-check that your JSON responses are being encoded with UTF-8 on the server side.
2. Use StandardCharsets.UTF_8 in Java:
If you're using Spring WebClient or similar frameworks, ensure that you set the correct character encoding. For example, in your WebClient configuration, you can specify the StandardCharsets.UTF_8 to ensure proper handling of special characters like emojis.
java
Copiar código
WebClient.builder()
    .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
    .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(16 * 1024 * 1024))
    .defaultCharset(StandardCharsets.UTF_8)  // Ensure UTF-8 encoding
    .build();
3. Custom Codec Configuration:
If you added your own custom codec, make sure that it's correctly registered and handling the UTF-8 encoding. Ensure that both the request and response bodies are being processed with the right encoding. You can use a LoggingCodec to inspect if the encoding is correctly set.
4. Check the Client-Side Rendering:
If the emojis are correctly encoded and sent by the server, but still not rendering properly on the client, it might be an issue with how the client is rendering the content. Make sure that your client (e.g., browser, mobile app, etc.) is set up to correctly display UTF-8 content. Ensure that the HTML page or the application is correctly configured to use UTF-8 by setting the charset in the HTML or response headers:
html
Copiar código
<meta charset="UTF-8">
5. Inspect the Response:
Use tools like Postman, curl, or browser developer tools to inspect the actual HTTP response and see if the emojis are present in the raw response body. This can help determine whether the issue is on the server side (encoding) or client side (rendering).
6. Dependency Conflicts:
Ensure that there are no conflicting dependencies or settings in your project that might override the intended behavior. Sometimes, other libraries or configurations can interfere with encoding.
7. Debug Logging:
Enable debug logging for WebClient or your HTTP client library to see detailed logs of the request/response lifecycle. This can provide insights into where the encoding might be going wrong.
8. Test in Isolation:
Create a simple test that sends and receives a JSON payload with emojis to verify that your WebClient configuration works as expected. This can help isolate the issue.
If none of these solutions work, you might want to share more details about the specific setup you're using, such as your WebClient configuration, server-side framework, or any custom codecs you've implemented.
